atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java

```diff
@@ -346,6 +346,23 @@ void setServersInPoolTo(ImmutableMap<CassandraServer, CassandraServerOrigin> des
                 .removeKeys(currentServers::contains)
                 .toImmutableMap();
         SetView<CassandraServer> absentServers = Sets.difference(currentServers, desiredServers.keySet());
+        Map<CassandraServer, CassandraClientPoolingContainer> currentPoolWithoutAbsentServers = EntryStream.of(
+                        getCurrentPools())
+                .removeKeys(absentServers::contains)
+                .toMap();
+
+        Set<CassandraServer> validatedServersToAdd =
+                validateNewHostsTopologiesAndMaybeAddToPool(currentPoolWithoutAbsentServers, serversToAdd);
```

Comment:
So there's technically a bit of new behaviour here, where `getCurrentPools()` can now return _more_ elements than what we believe the cluster to contain if a strategically placed read lands somewhere. `refreshPool` is synchronized on the CassandraClientPoolImpl, but the underlying `CassandraService#getPools` is not.

I did a check on where this map could be visible. This could technically lead to incorrect answers in `CassandraKeyValueServiceImpl#getNodeAvailabilityStatus()` because some nodes might get a double vote and we assess a quorum out of these. But this is anyway no worse than the behaviour today (since we can incorrectly conclude that we have quorum when there is none since we might give nodes a zero vote). We could fix this by having `getStatusByRunningOperationsOnEachHost()` acquire the lock and/or refactoring these modifications to have an AtomicReference style update...

But ultimately, because quorums for data manipulations happen at the Cassandra layer and are *not* reliant on what we do here, I think I'm ok with this.

Reply:
Hmm, maybe we then make `validateNewHostsTopologiesAndMaybeAddToPool` to be just `validateNewHostsTopologies`, and then we can remove then add, therefore keeping the behavior near-identical as it was before. I think we should be ok of the implications of the above, but that we don't accidentally expose new bugs.